src/find_service.rs

```diff
@@ -0,0 +1,151 @@
+use super::{ConnectionInfo, PublisherDesc};
+use futures::future::{self, Future};
+use futures::prelude::*;
+use futures::sync::oneshot::{self, Receiver, Sender};
+use futures::Async;
+use http;
+use hyper;
+use hyper::rt::Stream;
+use hyper::{client::HttpConnector, Body, Client, Uri};
+pub use io::Result;
+use serde_json;
+use std::result;
+use std::time;
+use tokio;
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct BlankResponse {}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct ConnectionResponse {
+    pub publisher: PublisherDesc,
+    pub info: ConnectionInfo,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct PubSubResponse<T> {
+    pub status: String,
+    pub timestamp: time::SystemTime,
+    pub response: T,
+}
+
+impl PubSubResponse<BlankResponse> {
+    pub fn status_only(status: String) -> PubSubResponse<BlankResponse> {
+        PubSubResponse {
+            status: status,
+            timestamp: time::SystemTime::now(),
+            response: BlankResponse {},
+        }
+    }
+}
+
+impl<T> PubSubResponse<T> {
+    pub fn new(status: String, response: T) -> PubSubResponse<T> {
+        PubSubResponse {
+            status: status,
+            timestamp: time::SystemTime::now(),
+            response: response,
+        }
+    }
+}
+#[derive(Serialize, Deserialize, Debug)]
+pub struct ServiceStatus {
+    pub count: usize,
+}
+
+type ServiceResponse = PubSubResponse<ServiceStatus>;
+
+pub enum ServerError {
+    Success,
+    HyperError(hyper::Error),
+    SerdeError(serde_json::Error),
+    UrlError(http::uri::InvalidUri),
+    StringError(String),
+}
+
+impl From<http::uri::InvalidUri> for ServerError {
+    fn from(uri_err: http::uri::InvalidUri) -> ServerError {
+        ServerError::UrlError(uri_err)
+    }
+}
+
+impl From<hyper::Error> for ServerError {
+    fn from(hyper_err: hyper::Error) -> ServerError {
+        ServerError::HyperError(hyper_err)
+    }
+}
+
+type FindClient = Client<HttpConnector, Body>;
+
+pub struct ResponseFuture<T> {
+    receiver: Receiver<result::Result<PubSubResponse<T>, ServerError>>,
+    error: ServerError,
+}
+
+impl<T> ResponseFuture<T> {
+    fn new() -> (
+        ResponseFuture<T>,
+        Sender<result::Result<PubSubResponse<T>, ServerError>>,
+    ) {
+        let (tx, rx) = oneshot::channel();
+        (
+            ResponseFuture {
+                receiver: rx,
+                error: ServerError::Success,
+            },
+            tx,
+        )
+    }
+}
+
+impl<'a, T> Future for ResponseFuture<T> {
+    type Item = PubSubResponse<T>;
+    type Error = ServerError;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
```

Comment:
wow you implemented your own future with poll. I feel like everything here should have been possible much more simply with map and map_err